### PR TITLE
ARTEMIS-3707 ResourceAdapter improvements

### PR DESCRIPTION
**Pull request by waldi5001**

### CODE REVIEW

**Comment by clebertsuconic** [comment]
Can you tell me a bit of background on this? why you're doing this?


These commits need a bit of squashing and more context on them...
What system did you integrate with in these? We need to check also on the integration with Wildfly. if you're integrating a new system it's totally fair but we need to make sure it does not break wildly as we already done it in there.

**Comment by waldi5001** [comment]
Hello @clebertsuconic, thanks for your reply. I wrote some words in the Jira ticket. 

I totally agree, not to break with wildly, but in my opinion, I heavn't. And if so, I will try fix it. 

It is no problem for me to squash or open a new pr for some of the commits. Have you a suggestion? 

Can you help me with the build? I can't see where is the problem.

**Comment by clebertsuconic** [comment]
@ehsavoie is there anyway you could run this branch against wildly integration?

Perhaps you both could collaborate on this?

**Comment by ehsavoie** [comment]
@clebertsuconic @waldi5001 I'll take a look to check it doesn't break with our current integration

**Comment by ehsavoie** [comment]
@waldi5001 I'm goig. Why did you remove    
   ActiveMQConnectionFactory getDefaultFactory() throws ResourceException 
   ActiveMQResourceAdapter getResourceAdapter() from ActiveMQRAConnectionFactory ? This is something that I use

**Comment by gemmellr** [comment]
There are lots of hard coded versions for dependencies and plugins in the poms. The versions should be set in properties and used via dependencyManagement and pluginManagement to convey the version to where needed. I think in many cases they will already be used that way elsewhere in the build, possibly with the same versions, which makes the hard coding even worse.

There look to be lots of tabbed indents introduced in the xml files instead of spaces, these should be fixed. In some cases existing spaces were converted to tabs and indents changed, creating a bunch of seemingly unnecessary change across fairly big files, the tabs should be fixed and it would be good to unwind unnecessary changes as the diff will be large enough as it is. Searching [the current diff](https://patch-diff.githubusercontent.com/raw/apache/activemq-artemis/pull/3999.diff) for tabs currently gives a lot of results.

**Comment by waldi5001** [comment]
> @waldi5001 I'm goig. Why did you remove ActiveMQConnectionFactory getDefaultFactory() throws ResourceException ActiveMQResourceAdapter getResourceAdapter() from ActiveMQRAConnectionFactory ? This is something that I use

If you want you can drop the commit 4fd3ef0d1554da6f43c4ec52e67c98b82b732687 "Polish", and than you can test further. Sorry, for me that code doesn't make sense.

**Comment by clebertsuconic** [comment]
merge commits are confusing..

Can you just 
```git pulll --rebase upstream main```

and also squash all these commits into one with an overall explanation of what you're doing as a single commit please?

**Comment by waldi5001** [comment]
> merge commits are confusing..
> 
> Can you just `git pulll --rebase upstream main`
> 
> and also squash all these commits into one with an overall explanation of what you're doing as a single commit please?

No problem. If the build is running and all is well, this will make sense.

**Comment by waldi5001** [comment]
@ehsavoie I've revert the commit. Can you recheck the wildfly integration?
@gemmellr I've reformat the files und centralize the versions. Can you review my the dependency management?

**Comment by clebertsuconic** [comment]
@waldi5001 why you keep using merge commits on a PR branch? :) just rebase against upstream/maini and have a single flow of commits?

**Comment by waldi5001** [comment]
@clebertsuconic The reason is: lack of knowledge :D. Now, I understand your comment
> merge commits are confusing..

I will avoid that.

should we split some commits in a new PR? Wouldn't that be cleaner.

File: docs/user-manual/en/resource-adapter.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+| | artemis-ra-rar   |     |     |
+| --------- | ------ | --- | --- |
+| Artemis   | JavaEE | JCA | JMS |
+| \>= 2.18.0 | \>=8      | 1.7 | 2.0 |
+| <= 2.17.0 | 7      | 1.5 | 2.0 |
+
+| | artemis-jakarta-ra-rar | | |
+| --- | ------- | --- | --- |
+| Artemis | JavaEE | JCA | JMS |
+| \>= 2.18.0 | \>=9 | 2.0 | 3.0 |
```
Comment (gemmellr) [inline docs/user-manual/en/resource-adapter.md:24]: The doc is version specific, I dont think it should concern itself with specifying stuff for what are even now already old Artemis versions. I would remove it or change it so it just describes the difference between the current rars, e.g

artemis-ra-rar A B C
artemis-jakarta-ra-rar X Y Z

File: docs/user-manual/en/resource-adapter.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+ResourceAdapter.
+
+## Details about the Resource Adapter configuration
+
+Before you start with the configuration you have to know two basics: The configuration is split into two 
+parts. First, the config to send messages to an address (outbound), and second, the config to get messages consumed 
+from a destination (inbound). Each can be configured separately, or use both the ResourceAdapter settings.
+
+Here are a few options listed. If you want an overview of all configuration options, consider
+[ConnectionFactoryProperties](https://github.com/apache/activemq-artemis/blob/main/artemis-ra/src/main/java/org/apache/activemq/artemis/ra/ConnectionFactoryProperties.java)
```
Comment (gemmellr) [inline docs/user-manual/en/resource-adapter.md:49]: Linking to main kind of sucks, and may not apply to whatever version is actually being looked at. Will whats wanted it be in the javadoc?

File: docs/user-manual/en/resource-adapter.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  
+  <jmsTopic jndiName="jms/sampleTopic">
+      <properties.artemis Address="a_sample_adress" />
+  </jmsTopic>
+  
+  <jmsQueue jndiName="jms/sampleQueue">
+      <properties.artemis Address="another_sample_adress" />
+  </jmsQueue>
+</server>
+```
```
Comment (gemmellr) [inline docs/user-manual/en/resource-adapter.md:130]: I dont see a need to embed this in the docs. It wont apply to many people and is just going to go stale. Seems like it would be better to just link to the example, which has it. Perhaps at the same time as mentioning other examples above.

File: examples/features/sub-modules/wlp/readme.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Apache ActiveMQ ResourceAdapter WLP / Openliberty Sample 
+
+This sample shows the usage of the Apache ActiveMQ Artemis ResourceAdapter with WLP / Openliberty.
+
+Usage:
+```bash
+mvn clean install
```
Comment (gemmellr) [inline examples/features/sub-modules/wlp/readme.md:7]: Seems the example doesnt even suggest using the previously commented about script.

File: pom.xml [hunk 5]
```diff
@@ -53,7 +53,9 @@
       <module>artemis-jakarta-server</module>
       <module>artemis-journal</module>
       <module>artemis-ra</module>
+      <module>artemis-ra-rar</module>
```
Comment (gemmellr) [inline pom.xml:56]: The doc additions speak about needing to build your own rar, and directs to the examples present for doing that. The 2 new modules being added at the root seem to be doing much the same thing. I'm not clear why both exist.

**Comment by clebertsuconic** [comment]
@waldi5001 

```
merge commits are confusing..
I will avoid that.

should we split some commits in a new PR? Wouldn't that be cleaner?
```


I would just for now:

```
git remote add upstream https://github.com/apache/activemq-artemis.git 
git pull --rebase upstream main
git push origin -f
```


Right before commit, I would squash all the commits into a single commit... it's ok to keep separate commits while reviewing.. but this to me is a single unit of work... 


if you want to separate commits to make it more manageable is fine.. but on this PR you have for instance a commit just to fix check style.. that should be squashed.


last time I checked all these commits could / should be squashed into a single commit and push -f

**Comment by clebertsuconic** [comment]
@waldi5001 look what I did at https://github.com/clebertsuconic/activemq-artemis/tree/ARTEMIS-3707


I basically got your branch.. rebased against main... squashed all the commits and pushed -f on my origin... you could do the same...


git is different from SVN on this sense..when you merge your work upstream, you can only merge commits as they make sense upstream.. we don't need every interaction you ever had with your code... it makes more confusing to follow on the history.



if you agree with the squash I made, you could type these commits:


```
git remote add clebert https://github.com/clebertsuconic/activemq-artemis.git
git fetch clebert
git checkout clebert/ARTEMIS-3707 -B ARTEMIS-3707
git push origin ARTEMIS-3707 -f
```


this should replace all your commits by my squashed.


if you want to squash them yourself.. first rebase your repo, fix a conflict you have on a commit and push -f as I showed you.


if you still having issues find me somewhere online and I can give you some help on this.

**Comment by waldi5001** [comment]
@gemmellr Thanks for your review. I will start to fix asap.
@clebertsuconic Thanks for you comprehensive help. I will test the solution and provide my work.

One question to both of you: In examples is shown to build your own .rar. With my commit Artemis provide a usable *.rar. Does it makes sense to keep the examples to build your own *.rar? The comment of @gemmellr here https://github.com/apache/activemq-artemis/pull/3999#discussion_r844151653 points in the same direction.

File: docs/user-manual/en/resource-adapter.md (hunk 2 above)
Comment (waldi5001) [reply docs/user-manual/en/resource-adapter.md:49]: of all your comments, this hurts the most :D In my opinion, link the src, is the best you could do, because the docu can't be outdated. With two mouseclicks you can select the tag version of the readers version and you are fine. Is link the src a no go for you? Or is it ok for you to add a hint, to select the right version of the src file?

File: pom.xml (hunk 5 above)
Comment (waldi5001) [reply pom.xml:56]: see my comment here https://github.com/apache/activemq-artemis/pull/3999#issuecomment-1090719349

File: docs/user-manual/en/resource-adapter.md (hunk 2 above)
Comment (gemmellr) [reply docs/user-manual/en/resource-adapter.md:49]: The docs are meant to be version specific, so linking to main for something that is documenting not-main, indeed kind of sucks. Sure, you can then change the version, but you can also fail to do so, which history suggests to me happens a whole bunch. We also aren't meant to direct regular everyday users to unreleased software as a matter of policy, which this somewhat skirts a little.

I asked about the javadoc, as it could then at least link to doc for the latest _released_ version, if not somehow the actual specific version (e.g I'd wonder if there wasnt some way to filter the version in), though I guess the latter would equally apply to the src.

**Comment by gemmellr** [comment]
> One question to both of you: In examples is shown to build your own .rar. With my commit Artemis provide a usable *.rar. Does it makes sense to keep the examples to build your own *.rar? The comment of @gemmellr here [#3999 (comment)](https://github.com/apache/activemq-artemis/pull/3999#discussion_r844151653) points in the same direction.

To be honest I was asking slightly more in the reverse direction of are these new modules necessary (after all this time), if you should still end up building your own which has multiple examples.

**Comment by waldi5001** [comment]
> To be honest I was asking slightly more in the reverse direction of are these new modules necessary (after all this time), if you should still end up building your own which has multiple examples.

Oh OK. My intention on this PR is / was: provide a *.rar file for everyone who want's to use Artemis with a resourceAdapter in a application server of their choice. I wanted to use Artemis myself with WLP and missed a ready to use *.rar like in ActiveMQ. So I startet my journey and I'm here :D.

My opinion: If you want to use Artemis with a application server that is not JBoss and you have to build your own *.rar, most of the user run away :D.

In other words: If Artemis provide a *.rar in the global mvn repo, no example is needed anymore and more people can use Artemis with the application server of their choice. :D

File: docs/user-manual/en/resource-adapter.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+| <= 2.17.0 | 7      | 1.5 | 2.0 |
+
+| | artemis-jakarta-ra-rar | | |
+| --- | ------- | --- | --- |
+| Artemis | JavaEE | JCA | JMS |
+| \>= 2.18.0 | \>=9 | 2.0 | 3.0 |
+
+## Lets start
+To use the RA, you have to build it by your own. This sounds harder than it is. But no worries, an 
+[examples](examples.md)
```
[the remark this answers is not in the document]
Comment (waldi5001) [reply docs/user-manual/en/resource-adapter.md:28]: I want to prove you completely right, but if you are looking the diff with "ignore whitespace" you can see it well. In my comment of the commit I describes the changes:
* update xml schemas
* fix ConnectionParameters
* remove TransactionManagerLocatorMethod and "The producer max rate"
* remove The max connections

nevertheless: With my next commit I reformat the file, and I hope the diff will become more readable.

File: examples/features/sub-modules/artemis-jakarta-ra-rar/src/main/resources/ra.xml [hunk 7]
```diff
@@ -2,314 +2,309 @@
 
 <!-- $Id: ra.xml 76819 2008-08-08 11:04:20Z jesper.pedersen $ -->
 
-<connector xmlns="http://java.sun.com/xml/ns/j2ee"
-           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-           xsi:schemaLocation="http://java.sun.com/xml/ns/j2ee
-           http://java.sun.com/xml/ns/j2ee/connector_1_5.xsd"
-           version="1.5">
+<connector version="2.0"
```
[the remark this answers is not in the document]
Comment (waldi5001) [reply examples/features/sub-modules/artemis-jakarta-ra-rar/src/main/resources/ra.xml:5]: I want to prove you completely right, but if you are looking the diff with "ignore whitespace" you can see it well. In my comment of the commit I describes the changes:
* update xml schemas
* fix ConnectionParameters
* remove TransactionManagerLocatorMethod and "The producer max rate"
* remove The max connections

nevertheless: With my next commit I reformat the file, and I hope the diff will become more readable.

**Comment by gemmellr** [comment]
> > To be honest I was asking slightly more in the reverse direction of are these new modules necessary (after all this time), if you should still end up building your own which has multiple examples.
> 
> Oh OK. My intention on this PR is / was: provide a *.rar file for everyone who want's to use Artemis with a resourceAdapter in a application server of their choice. I wanted to use Artemis myself with WLP and missed a ready to use *.rar like in ActiveMQ. So I startet my journey and I'm here :D.
> 
> My opinion: If you want to use Artemis with a application server that is not JBoss and you have to build your own *.rar, most of the user run away :D.
> 
> In other words: If Artemis provide a *.rar in the global mvn repo, no example is needed anymore and more people can use Artemis with the application server of their choice. :D

I see, none of that was clear to me as it doesnt seem to be obviously mentioned, and the PR even suggests the opposite in the new documentation added stating people needing to build their own, plus yet more example of that. At which point I questioned the seeming duplication; I don't actually use any of these bits so their usage is not familiar to me.

**Comment by waldi5001** [comment]
It was not my intention to secretly add a feature. If you take a look at the jira ticket, I've described, before the first commit my intentions. For me it was not clear whether a *rar would be accepted or not, and so it was also clear for me, not to align the documentary accordingly, but according to what was already there.

How can we continue? Or is it dead?

**Comment by gemmellr** [comment]
I did read the JIRA before reviewing, as I said that aspect wasnt (still isnt) clear to me, in part as I simply dont use these bits, but mainly due to the PR essentially contradicting itself with adding documentation saying you "must" build your own and another example doing that.

I dont think this is dead in the slightest. I do think there was a bunch of previous feedback you hadnt yet actioned (until perhaps now) even when repeated, or actually possibly still havent in the case of getting rid of the merge commits and squashing your changes to minimise the (now-getting-silly) number of commits, and rebasing things against the current main.

I havent had a chance to look at what you have now added since my previous comments yet, I will take another look when back from vacation next week.

File: examples/features/sub-modules/wlp-jakarta/pom.xml [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+   <artifactId>wlp-jakarta-resource-adapter-sample</artifactId>
+   <packaging>war</packaging>
+   <name>Artemis Websphere Liberty RA Jakarta Example</name>
+
+   <properties>
+      <activemq.basedir>${project.basedir}/../../../..</activemq.basedir>
+      <jakarta.jms-api.version>3.0.0</jakarta.jms-api.version>
+      <jakarta.annotation-api.version>2.0.0</jakarta.annotation-api.version>
+      <jakarta.ws.rs-api.version>3.0.0</jakarta.ws.rs-api.version>
+      <jakarta.ejb-api.version>4.0.0</jakarta.ejb-api.version>
```
Comment (gemmellr) [inline examples/features/sub-modules/wlp-jakarta/pom.xml:36]: Having every jakarta related module hard code a specific version of various specs will be quite annoying later. I have pushed some changes to main addressing this for existing stuff, see https://github.com/apache/activemq-artemis/commit/6eb78a26348db54164128c32c7321f1a8ae23c35. You should use the new props I added where appropriate, and follow the same approach to introduce other ones for the other deps this PR uses.

**Comment by gemmellr** [comment]
In addition to the specific comments, and even elaborating on a couple of them needing this, you really do need to rebase your changes against the current main and squash them. Clebert previously gave some advice and even did it for you at one point.

If you are still struggling doing that on the existing commits yourself you could just use the [**current diff generated by github**](https://patch-diff.githubusercontent.com/raw/apache/activemq-artemis/pull/3999.diff) from all your changes so far work to apply it to a fresh branch from main (then either update your current branch with that, or even open a new PR from the new branch). I was able to download the diff and apply it directly to the current main by just running `patch -p1 < 3999.diff`, with patch seeming to be able to handle all the divergences itself using fuzz or offsets.

**Comment by waldi5001** [comment]
After I have reconsidered the whole PR, I decided to split this big PR in three smaller one. I hope this is OK, even the last proposal was different.

https://github.com/apache/activemq-artemis/pull/4043 Remove of TransactionManager use
https://github.com/apache/activemq-artemis/pull/4044 Documentation of the ResourceAdapter
https://github.com/apache/activemq-artemis/pull/4045 Cleanup of the rar.xmls

If  PR https://github.com/apache/activemq-artemis/pull/4043 is merged, and there is hinteresst in a sample with WLP, I can provide it afterwards, in a new Jira Ticket and PR.

I close now this PR. Please give me feedback in the Jira Ticket [ARTEMIS-3707](https://issues.apache.org/jira/browse/ARTEMIS-3707) or also in the specific PR.
